### PR TITLE
boards: apollo3: Use helper function in main

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

Update to use a helper function in main to reduce the stack usage. This is to address the stack usage issue as described in https://github.com/tock/tock/issues/3518.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Comment by bradjc** [comment]
I'm not sure what this change does. It looks like the `main` stack frame is the same size:

```
From github.com:tock/tock
 * [new ref]             refs/pull/3520/head -> pr/3520
Switched to branch 'pr/3520'
❯ cd ..
❯ cd ../apollo3/lora_things_plus
❯ make stack-analysis
info: component 'llvm-tools' for target 'x86_64-apple-darwin' is up to date
lora_things_plus
----------------------
   main stack frame:
           24     main

   5 largest stack frames:
         2408     _ZN7cortexm24kernel_hardfault_arm_v7m17h88793f49ade3c395E
         1856     _ZN16lora_things_plus5setup17h49e91a52febf179eE
          832     _ZN95_$LT$kernel..process_standard..ProcessStandard$LT$C$GT$$u20$as$u20$kernel..process..Process$GT$18print_full_process17hd2851a4b78b89a34E
          592     _ZN6kernel5debug11panic_print17hf573d43131026b23E
          520     _ZN95_$LT$kernel..process_standard..ProcessStandard$LT$C$GT$$u20$as$u20$kernel..process..Process$GT$11try_restart17h699970668029a181E

❯ git co master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
❯ make stack-analysis
lora_things_plus
----------------------
   main stack frame:
           24     main

   5 largest stack frames:
         2408     cortexm::kernel_hardfault_arm_v7m
         1856     lora_things_plus::setup
          832     <kernel::process_standard::ProcessStandard<apollo3::chip::Apollo3<apollo3::chip::Apollo3DefaultPeripherals>> as kernel::process::Process>::print_full_process
          600     kernel::debug::panic_print::<lora_things_plus::io::Writer, apollo3::chip::Apollo3<apollo3::chip::Apollo3DefaultPeripherals>, kernel::process_printer::ProcessPrinterText>
          520     <kernel::process_standard::ProcessStandard<apollo3::chip::Apollo3<apollo3::chip::Apollo3DefaultPeripherals>> as kernel::process::Process>::try_restart
```

**Comment by alistair23** [comment]
The main difference is that we are now have `#[inline(never)]` on `setup()`. `setup()` might be inlined (as I guess it is in that comparison) but we now ensure it's never inlined.

File: chips/apollo3/src/lib.rs [hunk 1]
```diff
@@ -75,6 +75,7 @@ pub static IRQS: [unsafe extern "C" fn(); 32] = [CortexM4::GENERIC_ISR; 32];
 pub static PATCH: [unsafe extern "C" fn(); 16] = [unhandled_interrupt; 16];
 
 #[cfg(all(target_arch = "arm", target_os = "none"))]
+#[inline(always)]
```
Comment (bradjc) [inline chips/apollo3/src/lib.rs:78]: Why are we now inlining this?

Can you please comment to explain? We're learning that we can't keep track of why we are including inline directives on various functions, and we need to start documenting them.

Comment (alistair23) [reply chips/apollo3/src/lib.rs:80]: Comments added!

**Review by hudson-ayers — APPROVED** [review]

**Comment by hudson-ayers** [comment]
I think this PR is correct.

The two `#[inline(never)]`'s that were added are both on `setup()` calls to reduce stack frame size.

The added `#[inline(always)]` is definitely needed. If that call is not inlined, the board *will not run* because the `svc` call within relies on it being called from the main function. It not being there before was a bug, but Alistair probably did not hit the bug because it happened to be getting inlined in practice. This change might have caused LLVM to not inline the function, which would break things, so he probably noticed it as a result.

**Comment by bradjc** [comment]
Ok I somehow missed that we've been adding inline never with create_peripherals.

**Review by bradjc — APPROVED** [review]

**Pull request merged**